pkg/objectio/writer.go

```diff
@@ -0,0 +1,173 @@
+// Copyright 2021 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package objectio
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/fileservice"
+	"sync"
+)
+
+type ObjectWriter struct {
+	sync.RWMutex
+	object *Object
+	blocks map[int]BlockObject
+	buffer *ObjectBuffer
+	name   string
+	lastId int
+}
+
+func NewObjectWriter(name string, fs fileservice.FileService) (Writer, error) {
+	object := NewObject(name, fs)
+	writer := &ObjectWriter{
+		name:   name,
+		object: object,
+		buffer: NewObjectBuffer(name),
+		blocks: make(map[int]BlockObject),
+		lastId: 0,
+	}
+	err := writer.WriteHeader()
+	return writer, err
+}
+
+func (w *ObjectWriter) WriteHeader() error {
+	var (
+		err    error
+		header bytes.Buffer
+	)
+	h := Header{magic: Magic, version: Version}
+	if err = binary.Write(&header, binary.BigEndian, h.magic); err != nil {
+		return err
+	}
+	if err = binary.Write(&header, binary.BigEndian, h.version); err != nil {
+		return err
+	}
+	reserved := make([]byte, 22)
+	if err = binary.Write(&header, binary.BigEndian, reserved); err != nil {
+		return err
+	}
+	_, _, err = w.buffer.Write(header.Bytes())
+	return err
+}
+
+func (w *ObjectWriter) Write(batch *batch.Batch) (BlockObject, error) {
+	block := NewBlock(batch, w.object)
+	w.AddBlock(block.(*Block))
+	for i, vec := range batch.Vecs {
+		buf, err := vec.Show()
+		if err != nil {
+			return nil, err
+		}
+		offset, length, err := w.buffer.Write(buf)
+		if err != nil {
+			return nil, err
+		}
+		block.(*Block).columns[i].(*ColumnBlock).meta.location = Extent{
+			id:         block.GetMeta().header.blockId,
+			offset:     uint32(offset),
+			length:     uint32(length),
+			originSize: uint32(length),
+		}
+	}
+	return block, nil
+}
+
```

Comment:
What do you mean by Index here?   I am confused -- you mean bloom filter?

Reply:
write index has not yet been determined how to implement